File: from-0401-to-0500/0446-arithmetic-slices-ii---subsequence/solution.go
func numberOfArithmeticSlices(nums []int) int {
    n := len(nums)
    dp := make([]map[int]int, n)
    result := 0
    for i := 0; i < n; i ++ {
        dp[i] = make(map[int]int)
        for j := 0; j < i; j ++ {
            diff := nums[i] - nums[j]
            dp[i][diff] += (dp[j][diff] + 1)
            result += dp[j][diff]
        }
    }
    return result
}

File: from-0401-to-0500/0446-arithmetic-slices-ii---subsequence/solution.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        ans = 0
        n = len(nums)
        dp = [defaultdict(int) for _ in range(n)]
        for i in range(n):
            for j in range(i):
                diff = nums[i] - nums[j]
                dp[i][diff] += (dp[j][diff] + 1)
                ans += dp[j][diff]

        return ans